librad/src/meta/identity.rs

```diff
@@ -0,0 +1,1091 @@
+
+    #[error(transparent)]
+    Cjson(#[from] CjsonError),
+
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::error::Error),
+
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+/// Identity document ID
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct Revision(Oid);
+
+impl std::ops::Deref for Revision {
+    type Target = git2::Oid;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl AsRef<git2::Oid> for Revision {
+    fn as_ref(&self) -> &git2::Oid {
+        self
+    }
+}
+
+impl From<Revision> for git2::Oid {
+    fn from(rev: Revision) -> Self {
+        *rev.0.as_ref()
+    }
+}
+
+impl From<&Revision> for git2::Oid {
+    fn from(rev: &Revision) -> Self {
+        *rev.0.as_ref()
+    }
+}
+
+impl From<git2::Oid> for Revision {
+    fn from(oid: git2::Oid) -> Self {
+        Self(Oid(oid))
+    }
+}
+
+impl std::cmp::PartialOrd for Revision {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_bytes().partial_cmp(other.as_bytes())
+    }
+}
+
+impl std::cmp::Ord for Revision {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_bytes().cmp(other.as_bytes())
+    }
+}
+
+impl std::fmt::Display for Revision {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+/// Identity commit ID
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct ContentId(Oid);
+
+impl std::ops::Deref for ContentId {
+    type Target = git2::Oid;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl AsRef<git2::Oid> for ContentId {
+    fn as_ref(&self) -> &git2::Oid {
+        self
+    }
+}
+
+impl From<ContentId> for git2::Oid {
+    fn from(rev: ContentId) -> Self {
+        *rev.0.as_ref()
+    }
+}
+
+impl From<&ContentId> for git2::Oid {
+    fn from(rev: &ContentId) -> Self {
+        *rev.0.as_ref()
+    }
+}
+
+impl From<git2::Oid> for ContentId {
+    fn from(oid: git2::Oid) -> Self {
+        Self(Oid(oid))
+    }
+}
+
+impl std::fmt::Display for ContentId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::cmp::PartialOrd for ContentId {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_bytes().partial_cmp(other.as_bytes())
+    }
+}
+
+impl std::cmp::Ord for ContentId {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_bytes().cmp(other.as_bytes())
+    }
+}
+
+/// Delegations required to sign an identity document
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum Delegations {
+    /// Keys for the _user_ case
+    Keys(BTreeSet<PublicKey>),
+    /// Keys belonging to a user (identified by their `Revision`)
+    /// for the _project_ case
+    Users(BTreeMap<PublicKey, Revision>),
```

Comment:
> Why would a `User` document point to more `Users` as delegators?

The name can be misleading :-/

The `User` part is supposed to be used in project identities, where delegations are users.
The `Keys` part is a set of keys for a user identity.

So, yes, that fact that a user identity must use the `Keys` branch is not encoded in the types (same for the other case).

The problem is that the data needs to be deserialized anyway, there's nothing blocking buggy or malicious code from writing the wrong contents into a delegations set.

In this implementation the checks are there, and once the identity is verified you have the guarantee that the contents are correct.

Putting type arguments on the identity type will get us to the same problem we had for entities, where given a URN we could not know which type to use (for a user or for a project?), so we needed a union type anyway, and then we had to play tricks to make the union type use the _same_ serialized format as the generic type...

Reply:
> Why is the `PublicKey` the key here?

Because this is, ultimately, a set of keys, just like in the `Keys` case.
With the difference that each key "refers" to a user.